rfd/0131-adminitrative-actions-mfa.md

```diff
@@ -0,0 +1,274 @@
+- Adds a new system of passing user credentials, which can be used in other
+  endpoints that always or sometimes need MFA verification.
+  - Ex: `GenerateUserSingleUseCerts` could be replaced by simply checking for
+    MFA verification in the request metadata of `GenerateUserCerts`.
+- The MFA verification only applies to one API request, preventing MFA-replay-like
+  attacks.
+
+Cons:
+
+- Adding a new system of passing user credentials makes the client harder to
+  reason about, especially for external users of the API client.
+
+1. Reissue short-lived (1 minute TTL) MFA verified TLS certificates.
+
+This approach is similar to Per-session MFA. The client will need to issue a
+`GenerateUserSingleUseCerts` request to get MFA verified TLS certificates before
+making an administrative request.
+
+Pros:
+
+- Client can apply common logic for MFA verification, rather than endpoint
+  specific logic.
+  - Ex: First request without MFA, receive access denied error, re-request with MFA.
+  - MFA verification can be extended to additional Auth endpoint without any client changes.
+- Easily extends to other endpoints that always or sometimes need MFA verification.
+- Auth is completely handled by certificates rather than an augmented client request.
+
+Cons:
+
+- MFA verification applies to any requests using the MFA verified certificates.
+  This can be used as a MFA-replay-like attack since only one MFA verification is
+  required to carry out multiple administrative actions.
+  - Like Per-session MFA, this puts the security of the system more in the hands
+    of the client to not misuse or expose the certificates within their one
+    minute lifetime.
+- Requires users to create a new client with the reissued certificates, resulting
+  in additional connections to the Auth server.
+
+##### Option choice: TBD
+
+I am slightly in favor of option 2 due to the flexibility of the system compared
+to option 1 and the security benefit over option 3. However, we may favor the
+simplicity of defining MFA requirements on the gRPC layer (option 1) or of
+enforcing MFA based on the user's certificate alone (option 3).
+
+@Reviewers please offer your opinions and I will update this section once we
+decide on an option.
+
+#### Server changes
+
+For admin actions, the Auth server will validate MFA for each request, either
+from an `MFAChallengeResponse` passed by the client or an MFA-verified certificate
+used for the request.
+
+If the request fails MFA verification, an access denied
+error will be returned to the client.
+
+```go
+// ErrAPIMFARequired is returned by AccessChecker when an API request
+// requires an MFA check.
+var ErrAPIMFARequired = trace.AccessDenied("API request requires MFA")
+```
+
+The client will check for this error to determine whether it should retry a
+request with MFA verification.
+
+### Other considerations
+
+### Automated use cases
+
+If a user has no MFA device registered, they will not need to pass MFA
```

Comment:
I think if we provide an escape hatch for automated use cases, it can be miss used one way or another to bypass MFA. The important thing IMO is that in expected, safe usage of Teleport, an attacker cannot steal an admin's normal login session certs and wreak havoc on the cluster. If an admin is playing loose with automated user impersonation to avoid MFA, they are ultimately responsible for that. @jentfoo WDYT?

Also note, user impersonation itself will definitely require MFA.

Reply:
I think ideally we should start as restrictive as possible.  It's easier to make things more flexible as we understand customer needs than it is to make things more restrictive down the road.

That said, it's unclear to me if we can make this more restrictive without impacting the automated use cases.  So we probably just need to make sure this is well documented.  Perhaps we should add it to our administration security guide?

Let me know if I missed something, and there is another option here